cpp_linter/rest_api/github_api.py

```diff
@@ -446,6 +513,68 @@ def post_review(
                 summary_only=summary_only,
                 review_comments=review_comments,
             )
+        ignored_reviews = []
+        if not summary_only:
+            found_threads = self._get_existing_review_comments(
+                no_dismissed=reuse_comments and not delete_comments
+            )
+            if found_threads:
+                if reuse_comments:
+                    # Keep already posted comments if they match new ones
+                    review_comments_suggestions = review_comments.suggestions
+                    review_comments.suggestions = []
+                    existing_review_comments = []
+                    for thread in found_threads:
+                        for comment in thread["comments"]["nodes"]:
+                            found = False
+                            line_start = comment.get("originalStartLine", -1)
+                            line_end = comment["originalLine"]
+                            if comment["startLine"] is not None:
+                                line_start = comment.get(
+                                    "startLine", comment.get("line", line_start)
+                                )
+                                line_end = comment.get("line", line_end)
```

Comment:
I'm really just worried about `KeyError`s here.

If `"originalLine"` is guaraunteed to be in `comment`, then
```python
line_end = comment["originalLine"]
```
will not fail. Usually, we put `assert "originalLine" in comment` beforehand for typing assurance.

But still, you've got dead code here:
```python
                            if comment["startLine"] is not None:
                                line_start = comment.get(
                                    "startLine", comment.get("line", line_start)
                                )
```
If
```python
comment.get("startLine", None) == comment["startLine"]
```
then the code will be short-circuited to
```python
                            if comment.get("startLine", None) is not None:
                                line_start = comment["startLine"]
```

---
## Another swing
Am I getting close?
```suggestion
                            assert "originalLine" in comment, "GraphQL response malformed"
                            line_end = comment.get("line", comment.get("originalLine", -1))
                            line_start = comment.get("startLine", comment.get(
                                "line", comment.get("originalStartLine", line_end)
                            )
```

Reply:
Hhere are some examples of the graphql response for various review thread comments that can be used to help determine the correct logic. All of these are actually taken from this PR at time of posting:
```
{
              "id": "PRRT_kwDOHzaZPM5HWwCo",
              "isResolved": false,
              "isCollapsed": false,
              "comments": {
                "nodes": [
                  {
                    "id": "PRRC_kwDOHzaZPM5vh3vq",
                    "body": "```suggestion\r\n                            line_start = comment.get(\"originalStartLine\", -1)\r\n```\r\nThere's a lot of in-lined conditionals here. Most of them seem like they could be replaced with similar suggestions. There might need to be some added typing.",
                    "path": "cpp_linter/rest_api/github_api.py",
                    "line": null,
                    "startLine": null,
                    "originalLine": 467,
                    "originalStartLine": null
                  }
                ]
              }
            },
            {
              "id": "PRRT_kwDOHzaZPM5HWxv0",
              "isResolved": true,
              "isCollapsed": true,
              "comments": {
                "nodes": [
                  {
                    "id": "PRRC_kwDOHzaZPM5vh6Wa",
                    "body": "```suggestion\r\n    #: See :std:option:`--delete-review-comments`.\r\n    delete_comments: bool = True\r\n    #: See :std:option:`--reuse-review-comments`.\r\n    reuse_comments: bool = True\r\n```",
                    "path": "cpp_linter/cli.py",
                    "line": null,
                    "startLine": null,
                    "originalLine": 75,
                    "originalStartLine": 72
                  }
                ]
              }
            },
            {
              "id": "PRRT_kwDOHzaZPM5HaLjj",
              "isResolved": false,
              "isCollapsed": false,
              "comments": {
                "nodes": [
                  {
                    "id": "PRRC_kwDOHzaZPM5vnTaI",
                    "body": "I'm really just worried about `KeyError`s here.\r\n\r\nIf `\"originalLine\"` is guaranteed to be in `comment`, then\r\n```python\r\nline_end = comment[\"originalLine\"]\r\n```\r\nwill not fail. Usually, we put `assert \"originalLine\" in comment` beforehand for typing assurance.\r\n\r\nBut still, you've got dead code here:\r\n```python\r\n                            if comment[\"startLine\"] is not None:\r\n                                line_start = comment.get(\r\n                                    \"startLine\", comment.get(\"line\", line_start)\r\n
                     )\r\n```\r\nIf\r\n```python\r\ncomment.get(\"startLine\", None) == comment[\"startLine\"]\r\n```\r\nthen the code will be short-circuited to\r\n```python\r\n                            if comment.get(\"startLine\", None) is not None:\r\n                                line_start = comment[\"startLine\"]\r\n```\r\n\r\n---\r\n## Another swing\r\nAm I getting close?\r\n```suggestion\r\n                            assert \"originalLine\" in comment, \"GraphQL response malformed\"\r\n                            line_end = comment.get(\"line\", comment[\"originalLine\"])\r\n                            line_start = comment.get(\"originalStartLine\", line_end)\r\n                            if \"line\" in comment:\r\n                                line_start = comment.get(\"startLine\", comment[\"line\"])\r\n```",
                    "path": "cpp_linter/rest_api/github_api.py",
                    "line": 536,
                    "startLine": 530,
                    "originalLine": 536,
                    "originalStartLine": 530
                  }
                ]
              }
            },
```
The only example we dont have from this thread is where we have something along these lines:
```
                    "line": 465,
                    "startLine": null,
                    "originalLine": 467,
                    "originalStartLine": null,
```
So in order to match the output from the internally generated new comments to compare to them:
- for single line comments: `line_start`= -1, `line_end`="originalLine"
- for updated single line comments: `line_start`= -1, `line_end`="line"
- for multi line comments: `line_start`= "originalStartLine", `line_end`="originalLine"
- for updated multi line comments: `line_start`= "startLine", `line_end`="line"

So I have a redundant "startLine" check inside the conditional which if removed would make it:
```
line_start = comment.get("originalStartLine", -1)
line_end = comment["originalLine"]
if comment["startLine"] is not None:
    line_start = comment["startLine"]
    line_end = comment.get("line", line_end)
```
which would appear to not work for all the cases above (not sure why I did it that way unless there was some strange 5th case I dont remember)
So I would think it should be this:
```
line_end = comment.get("line", comment["originalLine"])
line_start = comment.get("startLine", comment.get("originalStartLine", -1))
```

Reply:
> ```python
> line_end = comment.get("line", comment["originalLine"])
> line_start = comment.get("startLine", comment.get("originalStartLine", -1))
> ```

Glad we worked through that logic. 👍🏼 This looks much cleaner, and it is closer to what I use when posting the review payloads.

Reply:
This tactic should also avoid using `null` or `0` values (line numbers are 1-base on github anyway)
```python
line_end = comment.get("line", None) or comment["originalLine"]
line_start = comment.get("startLine", None) or comment.get("originalStartLine", None) or -1
```
So
```json
{
  "line": 465,
  "startLine": null,
  "originalLine": 467,
  "originalStartLine": null
}
```
actually becomes
```python
{
  "line_start": -1,
  "line_end": 465,
}

```
whereas not using `None or default`, the above payload may be misinterpreted as
```python
{
  "line_start": None,
  "line_end": 465,
}
```

Reply:
Yep, I ran into that issue when testing locally, fixed and updated accordingly